### PR TITLE
[ Milestone 2 ] Decorate a class

### PR DESCRIPTION
**Pull request by CakeInTech**
- Created a class Nameable.
Implement a method called correct_name that will raise a NotImplementedError.
- Prepared base Decorator
- Prepared CapitalizeDecorator and TrimmerDecorator

### CODE REVIEW

File: nameable.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+class Nameable
+  def correct_name
+    raise NotImplementedError
+  end
+end
```
Comment (mikethreels) [inline nameable.rb]: - [ ] good job implementing the Nameable class, however, it seems like you either forgot to push or didn't create the remaining requirements listed below as well:

Turn your Person class to Nameable

    Make sure that your Person class inherits from Nameable
    Make sure that this class has a method correct_name implemented. It should simply return the name attribute.

Prepare base Decorator

    Make sure that it inherits from Nameable.
    In the constructor assign a nameable object from params to an instance variable.
    Implement the correct_name method that returns the result of the correct_name method of the @nameable.

Prepare CapitalizeDecorator and TrimmerDecorator

    For the CapitalizeDecorator:
        Create a class that inherits from the base Decorator class.
        Implement a method correct_name that capitalizes the output of @nameable.correct_name.
    For the TrimmerDecorator:
        Create a class that inherits from the base Decorator class.
        Implement a method correct_name that makes sure that the output of @nameable.correct_name has a maximum of 10 characters. If it's longer it should trim the word.

**Review by mikethreels — CHANGES_REQUESTED** [review]
Hi @MohamedCK, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights
- linters are passing

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: nameable.rb (hunk 1 above)
Comment (CakeInTech) [reply nameable.rb]: Hello @mikethreels , It seems like some of my commits disappeared. 😮

File: person.rb [hunk 2]
```diff
@@ -1,4 +1,11 @@
-class Person
+# rubocop:disable all
```
Comment (DammyShittu) [inline person.rb]: - [x] You disabled the rubocop linter completely. Kindly remove this disabling and resolve rubocop errors as this is to maintain standard code.

File: person.rb [hunk 3]
```diff
@@ -1,4 +1,11 @@
...
+class Nameable
+  def correct_name
+    raise NotImplementedError
+  end
+end
+
+class Person < Nameable
   attr_accessor :name, :age
   attr_reader :id
```
Comment (DammyShittu) [inline person.rb:10]: - [x] I got this error when I tried to test your decoration.

  image

  You can fix this by adding `super()` to the initialize method in the Person class and creating the missing `correct_name` method:

  ```
  class Person < Nameable
  
    def initialize(age, name = 'Unknown', parent_permission: true)
      ...
      super()
    end

    def correct_name
      @name
    end
  ```

File: person.rb [hunk 4]
```diff
@@ -23,3 +30,26 @@ def generate_id
...
+
+class Decorator < Nameable
+  def initialize(nameable)
+    @nameable = nameable
+  end
+
+  def correct_name
+    @nameable.correct_name
+  end
+end
```
Comment (DammyShittu) [inline person.rb]: - **[OPTIONAL]** Kindly move this class to its own file

File: person.rb [hunk 5]
```diff
@@ -23,3 +30,26 @@ def generate_id
...
+
+
+class CapitalizeDecorator < Decorator
+  def correct_name
+    @nameable.correct_name.capitalize
+  end
+end
```
Comment (DammyShittu) [inline person.rb:53]: - **[OPTIONAL]** Kindly move this class to its own file

File: person.rb [hunk 6]
```diff
@@ -23,3 +30,26 @@ def generate_id
...
+
+class TrimmerDecorator < Decorator
+  def correct_name
+    @nameable.correct_name[0..9]
+  end
+end
```
Comment (DammyShittu) [inline person.rb]: - **[OPTIONAL]** Kindly move this class to its own file

File: person.rb [hunk 7]
```diff
@@ -1,4 +1,11 @@
-class Person
+# rubocop:disable all
+class Nameable
+  def correct_name
+    raise NotImplementedError
+  end
+end
```
Comment (DammyShittu) [inline person.rb:5]: - **[OPTIONAL]** Kindly move this class to its own file

**Review by DammyShittu — CHANGES_REQUESTED** [review]
Hi @MohamedCK,

## Highlights
✔️ Descriptive PR
✔️ Passing linters
✔️ Good branch name

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: person.rb (hunk 4 above)
Comment (CakeInTech) [reply person.rb]: @DammyShittu  I did that from the start but it doesn't work even tho I used the `require_relative` to import the `nameable` file so I'm having trouple there.

File: person.rb (hunk 2 above)
Comment (CakeInTech) [reply person.rb]: I have turned it on and the error it's solved thanks! 😄

File: person.rb (hunk 3 above)
Comment (CakeInTech) [reply person.rb:10]: I did as you suggested by adding the super method  and also the correct_name method thanks. 😄

File: person.rb (hunk 5 above)
Comment (CakeInTech) [reply person.rb:53]: @DammyShittu I did that from the start but it doesn't work even tho I used the require_relative to import the nameable file so I'm having trouple there.

File: person.rb (hunk 6 above)
Comment (CakeInTech) [reply person.rb]: @DammyShittu I did that from the start but it doesn't work even tho I used the require_relative to import the nameable file so I'm having trouple there.

File: person.rb (hunk 7 above)
Comment (CakeInTech) [reply person.rb:5]: @DammyShittu I did that from the start but it doesn't work even tho I used the require_relative to import the nameable file so I'm having trouple there.

**Comment by CakeInTech** [comment]
Hi @DammyShittu , 

Thank you for the optional messages but I wasn't able to put each code line to it's own file so hopefully you might be able to help me with that thanks. 😄

**Review by gemmen29 — APPROVED** [review]
Hi @MohamedCK, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights 💯 
- No linters errors ✔️ 
- Descriptive pull request ✔️ 
- Descriptive commit messages ✔️ 



## Optional suggestions

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**